library/common/http/dispatcher.cc

```diff
@@ -49,75 +65,152 @@ void Dispatcher::DirectStreamCallbacks::onHeaders(HeaderMapPtr&& headers, bool e
+    bridge_callbacks_.on_error({code, message}, bridge_callbacks_.context);
+  }
+  // FIXME: before merge. this can't be right because then we could be potentially calling cleanup
+  // twice on a stream. For instance on an inline reset? Not for the case with the local_closed_
+  // false but remote_closed_ true. Because in that case the cleanup call is guarded by complete(),
+  // right?
+  http_dispatcher_.cleanup(direct_stream_.stream_handle_);
+}
+
+void Dispatcher::DirectStreamCallbacks::onCancel() {
+  // This call is guarded at the call-site @see Dispatcher::DirectStream::resetStream().
+  // Therefore, it is dispatched here without protection.
+  ENVOY_LOG(debug, "[S{}] dispatching to platform cancel stream", direct_stream_.stream_handle_);
+  bridge_callbacks_.on_cancel(bridge_callbacks_.context);
+
+  // FIXME: before merge. what about clean up? It all depends on if the reset we send down causes a
+  // reset coming up, which will not fire a callback, but will result in stream clean up.
+}
+
+Dispatcher::DirectStream::DirectStream(envoy_stream_t stream_handle, Dispatcher& http_dispatcher)
+    : stream_handle_(stream_handle), parent_(http_dispatcher) {}
+
+void Dispatcher::DirectStream::resetStream(StreamResetReason) { callbacks_->onReset(); }
+
+void Dispatcher::DirectStream::closeLocal(bool end_stream) {
+  // TODO: potentially guard against double local closure.
+  local_closed_ = end_stream;
+
+  // No cleanup happens here because cleanup always happens on remote closure.
+  // see @Dispatcher::DirectStreamCallbacks::closeRemote.
+  // Remote closure is guaranteed to happen either through the golden path, or through an Envoy
+  // driven reset.
+}
+void Dispatcher::DirectStream::closeRemote(bool end_stream) {
+  remote_closed_ = end_stream;
+  // Even though we have potentially fired an on_complete callback to the bridge_callbacks,
+  // We will only cleanup if **both** the local and the remote have closed because otherwise we
+  // can be expecting a reset from envoy.
+  // See the comment for @see Dispatcher::DirectStreamCallbacks::closeRemote().
+  if ((complete())) {
+    ENVOY_LOG(debug, "[S{}] scheduling cleanup", stream_handle_);
+    parent_.cleanup(stream_handle_);
+  }
 }
 
-Dispatcher::DirectStream::DirectStream(envoy_stream_t stream_handle,
-                                       AsyncClient::Stream& underlying_stream,
-                                       DirectStreamCallbacksPtr&& callbacks)
-    : stream_handle_(stream_handle), underlying_stream_(underlying_stream),
-      callbacks_(std::move(callbacks)) {}
-
-AsyncClient::StreamOptions
-Dispatcher::DirectStream::toNativeStreamOptions(envoy_stream_options stream_options) {
-  AsyncClient::StreamOptions native_stream_options;
-  native_stream_options.setBufferBodyForRetry(stream_options.buffer_body_for_retry);
-  return native_stream_options;
+bool Dispatcher::DirectStream::complete() { return local_closed_ && remote_closed_; }
```

Comment:
Is this true? Do we always set `local_closed_`?

Reply:
no, but that is handled by the reset that envoy would cause. In addition to the threading of cancellation, this (terminal states) is the other thing I'd like to talk through.